src/main/java/com/vinted/flink/bigquery/sink/defaultStream/BigQueryDefaultSinkWriter.java

```diff
@@ -107,32 +97,70 @@ public AppendCallBack(BigQueryDefaultSinkWriter<A, ?> parent, String traceId, Ro
-                var second = rows.getData().subList(rows.getData().size() / 2, rows.getData().size());
-                try {
-                    this.parent.writeWithRetry(traceId, rows.updateBatch(first, rows.getOffset()), retryCount);
-                    this.parent.writeWithRetry(traceId, rows.updateBatch(second, rows.getOffset() + first.size()), retryCount);
-                } catch (Throwable e) {
-                    this.parent.appendAsyncException = new AppendException(traceId, rows, t);
+            try {
+                if (newRetryCount > 0) {
+                    logger.warn("Trace-id {} Recoverable error {}. Retrying...", traceId, status.getCode(), t);
+                    this.parent.writeWithRetry(traceId, rows, newRetryCount);
+                } else {
+                    logger.error("Trace-id {} Recoverable error {}. No more retries left", traceId, status.getCode(), t);
+                    this.parent.appendAsyncException = new AppendException(traceId, rows, newRetryCount, t);
                 }
-            } else {
-                this.parent.appendAsyncException = new AppendException(traceId, rows, t);
+            } catch (Throwable e) {
+                this.parent.appendAsyncException = new AppendException(traceId, rows, newRetryCount, e);
+            }
+        }
+
+        private void doPauseBeforeRetry() {
+            try {
+                Thread.sleep(parent.clientProvider.writeSettings().getRetryPause().toMillis());
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
             }
         }
 
         @Override
-        public void onSuccess(AppendRowsResponse result) {
-            Optional.ofNullable(this.parent.metrics.get(rows.getStream())).ifPresent(m -> {
-                m.setBatchCount(rows.getData().size());
-                m.setOffset(result.getAppendResult().getOffset().getValue());
-            });
+        public void onFailure(Throwable t) {
+            logger.error("Trace-id {} Received error {}", t.getMessage(), traceId);
+            var status = Status.fromThrowable(t);
+            switch (status.getCode()) {
+                case INTERNAL:
+                case ABORTED:
+                case CANCELLED:
+                case FAILED_PRECONDITION:
+                case DEADLINE_EXCEEDED:
+                case UNAVAILABLE:
+                    doPauseBeforeRetry();
+                    retryWrite(t, retryCount - 1);
+                    break;
+                case INVALID_ARGUMENT:
+                    if (t.getMessage().contains("INVALID_ARGUMENT: MessageSize is too large.")) {
+                        Optional.ofNullable(this.parent.metrics.get(rows.getStream())).ifPresent(BigQueryStreamMetrics::incSplitCount);
+                        logger.warn("Trace-id {} MessageSize is too large. Splitting batch", traceId);
+                        var first = rows.getData().subList(0, rows.getData().size() / 2);
```

Comment:
could we have problems here if size is not an even number?

Reply:
extracted to var, there's no problem if size is not event number